### ISSUE TITLE
NeighborSampler connected target nodes and their n_id

### ISSUE THREAD

**Issue by raphaelsulzer** [rusty1s/pytorch_geometric#1393]
## ❓ Questions & Help

Hi,

is there any way to force the neighbor sampler to have (1-hop) connected target nodes for the last layer?

The reason is that I would like to implement a pairwise regularization loss for my network.

I am also confused about the n_id attribute of the neighbor sampler. In the documentation you say that they are: "the IDs n_id of all nodes involved in the computation". However, size(n_id) is for me always the same as size adj.size[0], meaning n_id only holds the source nodes of the first layer, or am I wrong? How would I get the n_ids of the target nodes of my final layer for example?

**Comment by rusty1s** [comment]
The target nodes are also included in n_id (and are placed first in n_id. The target node id of the last layer hence corresponds to n_id[:batch_size].

**Comment by raphaelsulzer** [comment]
Thanks for the quick reply!

I want my target nodes (of the last layer) to be one connected component of my large graph, instead of just randomly distributed nodes that are not connected to each other. This will of course lead to duplicates in the n_ids etc. Could you give me a hint on how to achieve this?

**Comment by rusty1s** [comment]
I think the easiest way to achieve that would be to modify the NeighborSampler class directly. Instead of passing randomly shuffled node indices to the underlying DataLoader, you can pass in a list of a list of connected components, and use that in conjunction with a batch size of 1.

**Comment by raphaelsulzer** [comment]
ok, thank you! I will give that a try and report back if it worked...

**Comment by raphaelsulzer** [comment]
I tried to do what you said, by passing a custom torch.utils.data.Sampler to the NeighborSampler via the sampler arg (which is in turn passed to the underlying torch.utils.data.DataLoader). However, for some reason it overwrites the sampler function in the NeighborSampler (which is passed via the collate_fn arg to the underlying DataLoader). I couldn't quite figure out why this happens.

Long story short, I achieved what I wanted to do by simply sampling more hops than I have layers in my network. This also leaves me with final embeddings for connected components...
